wagtail/admin/filters.py

```diff
@@ -67,6 +68,128 @@ def decompress(self, value):
         return [None, None]
 
 
+class FilteredSelect(forms.Select):
+    """
+    A select box variant that adds 'data-' attributes to the <select> and <option> elements
+    to allow the options to be dynamically filtered by another select box.
+    See wagtailadmin/js/filtered-select.js for an example of how these attributes are configured.
+    """
+
+    def __init__(self, attrs=None, choices=(), filter_field=''):
+        super().__init__(attrs, choices)
+        self.filter_field = filter_field
+
+    def build_attrs(self, base_attrs, extra_attrs=None):
+        my_attrs = {
+            'data-widget': 'filtered-select',
+            'data-filter-field': self.filter_field,
+        }
+        if extra_attrs:
+            my_attrs.update(extra_attrs)
+
+        return super().build_attrs(base_attrs, my_attrs)
+
+    def optgroups(self, name, value, attrs=None):
+        # copy of Django's Select.optgroups, modified to accept filter_value as a
+        # third item in the tuple and expose that as a data-filter-value attribute
+        # on the final <option>
+        groups = []
+        has_selected = False
+
+        for index, choice in enumerate(self.choices):
+            try:
+                (option_value, option_label, filter_value) = choice
+            except ValueError:
+                # *ChoiceField will still output blank options as a 2-tuple,
+                # so need to handle that too
+                (option_value, option_label) = choice
+                filter_value = None
+
+            if option_value is None:
+                option_value = ''
+
+            subgroup = []
+            if isinstance(option_label, (list, tuple)):
+                group_name = option_value
+                subindex = 0
+                choices = option_label
+            else:
+                group_name = None
+                subindex = None
+                choices = [(option_value, option_label)]
```

Comment:
It looks like sub-choices can't have a separate filter value to the rest of the group. Is this intentional then?

Reply:
I didn't consider how this would interact with grouped options - the logic for dealing with them is only here because the base Select implementation funnels everything through the `optgroups` method, and I wanted to make the changes to the original method as minimal as possible. As it stands, I think you're right - the effect of that is that we only accept a filter_value at the top level of `choices` (i.e. an optgroup or ungrouped choice). Can look at supporting filter_value within optgroups too, if you think that's a useful addition?

Reply:
I don't think it's a massively important feature, but it would be nice to support it if it's not too much effort, just for continuity. Otherwise I'd suggest just making a note in comment or docstring that this isn't currently supported, as I can see it being a bit confusing to narrow that down - I'd be happy with either!

Reply:
optgroup support added in 3f18f0f8a2aa4c1f6859ae1adfabde5cdf90423d